File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.swift
class Solution {
    
    var upper = -1, lower = -1

    
      func findBound( _ nums: [Int], _ lo: Int, _ hi: Int, _ target: Int,_ isFirst: Bool){
        var lo = lo, hi = hi
        while( lo <= hi){
            var mid = (lo + hi) / 2
            
            if nums[mid] == target{
                if isFirst{
                    lower = mid 
                    hi = mid - 1
                }else{
                    upper = mid 
                    lo  = mid + 1
                }
                
            }else if nums[mid] > target{
                hi = mid - 1
            }else{
                lo = mid + 1
            }
            
            
        }
    }
    
    func searchRange(_ nums: [Int], _ target: Int) -> [Int] {
        findBound(nums, 0, nums.count - 1, target, true)
         findBound(nums, 0, nums.count - 1, target, false)
    
        var ans = [lower, upper]

        return ans
        
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int binarySearch(int lo , int hi , int num , vector<int>& nums , bool flag){
        int mid , ans = -1;
        while( lo <= hi ){
            mid = lo + ( hi - lo )/2;
            if ( nums[mid] > num ){
                hi = mid -1 ;
            }else if ( nums[mid] < num){
                lo = mid + 1 ;
            }else {
                ans = mid ;
                if ( flag == false){
                     hi = mid - 1 ;
                }else {
                     lo = mid + 1;
                }
               
            }
        }
        return ans;
    }
    
    
    vector<int> searchRange(vector<int>& nums, int target) {
        int lower_bound = binarySearch(0,nums.size()-1,target,nums,false);
        int upper_bound = binarySearch(0,nums.size()-1,target,nums,true);
        vector < int > ans;
        ans.push_back(lower_bound);
        ans.push_back(upper_bound);
        return ans;
        
    }
};